### DISCUSSION TITLE
Modifying TSNode

### DISCUSSION THREAD

**Discussion by nhasabni** [tree-sitter/tree-sitter#1046] [Q&A] [answered]
Hi,

I'm interested to know if there is easier and robust way to capture certain information that could be needed for program analysis after building ASTs. For instance, I would like to capture variable names in TSNode objects for identifiers. Such information could be important for program analysis using parse trees. I can add such information by modifying in-memory TSNode objects, but I am not sure if this is robust approach. Is there any better way to achieve this?

Also, are there OCaml bindings for TSNode objects (I found Haskell support, but not OCaml)? I am interested in performing program analysis on TS parse trees in OCaml.

Thank you.

**Comment by ahlinc** [comment]
https://github.com/returntocorp/ocaml-tree-sitter

**Reply by nhasabni** [reply]
Thanks. Any thoughts on modifying TSNode objects?

**Reply by nhasabni** [reply]
Also, the link that you mentioned is using GPL license, while Tree-Sitter is using MIT license. Not sure how the two work together.

**Reply by ahlinc** [reply, marked as answer]
I've posted a link to most maintainable OCaml binding that I know, regarding the licence I also doesn't know how it's compatible to this repo.

It's unclear for me why do you want to modify TSNode objects. For your example you can capture all corresponding nodes by writing a suitable tree-sitter query.
